### PR TITLE
make count(f,itr) count the number of nonzero; countnz(itr) -> count(itr)

### LINKED ISSUE #20403: count(itr) to count trues
I think it would be useful to add a method of `count` that takes a single iterable argument of booleans and returns the number of `true` elements.

For example (imagine that the function `iseven` does not exist), I find this:

```julia
count(i % 2 == 0 for i in 1:10)
```

to be more readable than this:

```julia
count(i -> i % 2 == 0, 1:10)
```

### PR DESCRIPTION
**Pull request by stevengj**
Closes #20403, #20404.

### CODE REVIEW

File: base/reduce.jl [hunk 1]
```diff
@@ -637,42 +637,37 @@ function contains(eq::Function, itr, x)
...
 
-julia> countnz(A)
+julia> count(A)
 6
 ```
 """
-countnz(a) = count(x -> x != 0, a)
+function count end
+
+function count{F}(pred::F, itr)
```
Comment (ararslan) [inline base/reduce.jl]: The static parameter shouldn't be necessary I would think

Comment (stevengj) [inline base/reduce.jl]: It used to be necessary in order to tell the compiler to type-specialize this on function arguments, I thought.  Has that changed?

Comment (ararslan) [inline base/reduce.jl]: I thought so but I'm probably wrong. @vtjnash?

Comment (stevengj) [inline base/reduce.jl]: Hmm, I just tried it, and it seems to inline fine without the parameters now.

Comment (Sacha0) [inline base/reduce.jl]: IIRC, calling a function argument within the method body triggers specialization on the function argument (hence forcing specialization on `pred` via a static parameter is unnecessary in this case). Ref. #19137. Best!

File: base/reduce.jl [hunk 2]
```diff
@@ -637,42 +637,37 @@ function contains(eq::Function, itr, x)
...
 
-
-## countnz & count
-
 """
     count(p, itr) -> Integer
+    count(itr)
 
-Count the number of elements in `itr` for which predicate `p` returns `true`.
+Count the number of elements in `itr` for which the function `p` returns
```
Comment (nalimilan) [inline base/reduce.jl]: Would make sense to update `find` to use `iszero` too in a subsequent commit.

File: base/reduce.jl [hunk 3]
```diff
@@ -637,42 +637,37 @@ function contains(eq::Function, itr, x)
     return false
 end
 
-
-## countnz & count
-
 """
     count(p, itr) -> Integer
+    count(itr)
```
Comment (nalimilan) [inline base/reduce.jl]: Repeat `-> Integer`? BTW, should be written as `::Integer` now.

File: base/reduce.jl [hunk 4]
```diff
@@ -637,42 +637,37 @@ function contains(eq::Function, itr, x)
...
+julia> count(A)
 6
 ```
 """
-countnz(a) = count(x -> x != 0, a)
+function count end
+
+function count(pred, itr)
+    n = 0
+    for x in itr
```
Comment (nalimilan) [inline base/reduce.jl]: Not really related, but should we have a more efficient version for `AbstractArray` which would add `@inbounds`? When `pred=identity` in particular, a SIMD version should be much more efficient.

File: base/reduce.jl (hunk 3 above)
Comment (ararslan) [inline base/reduce.jl]: I don't think most of the Base docstrings have made the `::` switch yet. IMO better to wait and do them all at once, keeping `->` in the interim.

**Comment by stevengj** [comment]
Closing in favor of #20421 since the `countnz` changes are controversial.

**Pull request closed without merging**